File: frontend/src/types/embedding.ts
/**
 * app/schemas/embedding.py -> TypeScript Interfaces
 */

// --- Backend Schema Interfaces ---

export interface EmbeddingBlockLite {
  accession_no?: string | null;
  specimen_label?: string | null;
  block_no?: number | null;
  block_code?: string | null;
  is_decal?: boolean | null;
}

export interface EmbeddingDetailResponse {
  id: number;
  run_id: number;
  block_id: number;
  embedded_at: string; // ISO DateTime string
  block?: EmbeddingBlockLite | null;
}

export interface EmbeddingRunResponse {
  id: number;
  run_no: string;
  started_at: string; // ISO DateTime string
  user_id: number;
  user_full_name?: string | null;
  station_id?: string | null;
  details?: EmbeddingDetailResponse[]; // ปรับเป็น Optional เผื่อกรณี List หน้าแรกไม่ส่งรายละเอียดมา
}

export interface EmbeddingRunCreate {
  user_id: number;
  station_id?: string;
}

export interface ScanBlockRequest {
  run_id: number;
  block_no: string; // บาร์โค้ดที่สแกนมา (เช่น A1 หรือ Barcode ID)
}

// --- Interfaces สำหรับ Pending Data (Modal Tree) ---

export interface PendingBlock {
  key: number; // id ของ block
  id: number;
  code: string; // block_code เช่น "A1"
  isCase: boolean; // false
  is_decal: boolean;
  accession_no: string;
}

export interface PendingTree {
  key: string; // case-{id}
  id: number;
  code: string; // accession_no เช่น "S26-00001"
  isCase: boolean; // true
  children: PendingBlock[]; // สำหรับ Ant Design Table Tree
}

// --- Component Props Interfaces ---

export interface EmbeddingRunListProps {
  onCreateClick: () => void;
  onSelectRun: (id: number) => void;
  refreshKey?: number;
}

export interface CreateEmbeddingRunProps {
  onBack: () => void;
}

// --- Local UI Types ---

export type EmbeddingViewMode = "list" | "create" | "details";

File: backend/app/schemas/embedding.py
# app/schemas/embedding.py
from pydantic import ConfigDict, BaseModel, model_validator
from datetime import datetime
from typing import Any, List, Optional

class EmbeddingBlockLite(BaseModel):
    accession_no: Optional[str] = None
    specimen_label: Optional[str] = None
    block_no: Optional[int] = None
    block_code: Optional[str] = None
    is_decal: Optional[bool] = None

    model_config = ConfigDict(from_attributes=True)

class EmbeddingDetailResponse(BaseModel):
    id: int
    run_id: int
    block_id: int
    embedded_at: datetime
    block: Optional[EmbeddingBlockLite] = None

    model_config = ConfigDict(from_attributes=True)

class EmbeddingRunResponse(BaseModel):
    id: int
    run_no: str
    started_at: datetime
    user_id: int
    user_full_name: Optional[str] = None
    details: List[EmbeddingDetailResponse] = []

    @model_validator(mode="before")
    @classmethod
    def populate_user_full_name(cls, data: Any) -> Any:
        if hasattr(data, "user") and data.user:
            data.__dict__["user_full_name"] = data.user.full_name or data.user.username
        return data

    model_config = ConfigDict(from_attributes=True)
        
class EmbeddingRunCreate(BaseModel):
    user_id: int
    station_id: Optional[str] = None

class ScanBlockRequest(BaseModel):
    run_id: int
    block_no: str # รับเป็นชื่อบาร์โค้ดที่สแกนมา